### ISSUE TITLE
deis builder often leaves stale node modules for nodejs Heroku buildpack

### ISSUE THREAD

**Issue by jbeard4** [deis/deis#3652]
I have a node module which uses git URLs for some dependencies:

```json
slug@85833f869081:~$ cat package.json 
{
  "name": "scion-sandbox-worker",
  "version": "0.0.0",
  "private": true,
  "scripts": {
    "start": "node ./bin/www"
  },
  "dependencies": {
    "body-parser": "~1.10.2",
    "cookie-parser": "~1.3.3",
    "debug": "^2.1.3",
    "express": "~4.11.1",
    "jade": "~1.9.1",
    "knox": "^0.9.2",
    "morgan": "~1.5.1",
    "redis": "~0.12.1",
    "request": "^2.55.0",
    "scxml": "git://github.com/jbeard4/SCION.git#44436d1a9294dc32c358cb84d3e1d0400dae0c4f",
    "serve-favicon": "~2.2.0",
    "underscore": "^1.8.3"
  }
}
```

You can see that the actual version of `scxml` module in the slug has the incorrect SHA1: `5a85d4a2d948d7792c6ccb73a35201c412c78652`

```json
slug@85833f869081:~/node_modules/scxml$ cat package.json 
{
  "name": "scxml",
  "version": "2.0.1",
  "description": "An implementation of SCXML in JavaScript.",
  "keywords": [
    "scxml",
    "statecharts",
    "w3c",
    "javascript"
  ],
  "maintainers": [
    {
      "name": "jbeard",
      "email": "jbeard4@cs.mcgill.ca"
    }
  ],
  "bugs": {
    "url": "https://github.com/jbeard4/scxml.js/issues",
    "email": "jbeard4@cs.mcgill.ca"
  },
  "licenses": [
    {
      "type": "Apache-2.0",
      "url": "https://github.com/jbeard4/scxml.js/blob/master/LICENSE.txt"
    }
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/jbeard4/scxml.js.git"
  },
  "implements": [
    "http://www.w3.org/TR/scxml/"
  ],
  "dependencies": {
    "optimist": "0.6.1",
    "sax": "0.6.0",
    "scion-core": "1.0.0",
    "esprima": "~1.2.2"
  },
  "devDependencies": {
    "karma": "~0.8"
  },
  "engines": {
    "node": ">=0.8",
    "npm": "1.4.10"
  },
  "main": "./lib/runtime/platform-bootstrap/node/index.js",
  "scripts": {
    "test": "cd test && bash run-tests-node.sh"
  },
  "bin": {
    "scxml": "./bin/repl.js"
  },
  "gitHead": "5a85d4a2d948d7792c6ccb73a35201c412c78652",
  "homepage": "https://github.com/jbeard4/scxml.js",
  "_id": "scxml@2.0.1",
  "_shasum": "ca9c0586840447b39cccb48627539afd2f306e67",
  "_from": "scxml@>=2.0.1 <3.0.0",
  "_npmVersion": "1.4.28",
  "_npmUser": {
    "name": "jbeard",
    "email": "jbeard4@cs.mcgill.ca"
  },
  "dist": {
    "shasum": "ca9c0586840447b39cccb48627539afd2f306e67",
    "tarball": "http://registry.npmjs.org/scxml/-/scxml-2.0.1.tgz"
  },
  "directories": {},
  "_resolved": "https://registry.npmjs.org/scxml/-/scxml-2.0.1.tgz"
}
```

I believe the deis builder is failing to fetch the latest version of the scxml submodule, leaving stale node_modules. I have seen this happen in other projects.

I've tried to set Heroku's NODE_MODULES_CACHE=false, but I'm not sure whether this is supported by deis: https://devcenter.heroku.com/articles/nodejs-support#cache-behavior

The only workaround I have found is to delete the module in package.json, and then re-add it. This forces the builder to download the correct module. However, this requires two distinct pushes to deis, doubling build time, and leaves bogus commits in my git history, so it's not an ideal solution.

**Comment by bacongobbler** [comment]
This sounds like a buildpack issue. NODE_MODULES_CACHE is used [by the buildpack](https://github.com/heroku/heroku-buildpack-nodejs/blob/d3952c3ef928cb63066e9306560b4fd56c8aebdf/lib/build.sh#L94), not the platform. It is possible that the build environment could be setting up environment variables improperly. however. We have hit issues like that in the past: #3212

**Comment by jbeard4** [comment]
I think it might not be setting env vars correctly for the buildpack:

```
jacob@jacob-ThinkPad-W520:~/workspace/jacobean/scion-sandbox-worker$ deis config:list
=== scion-sandbox-worker Config
NODE_MODULES_CACHE      false
```

```
jacob@jacob-ThinkPad-W520:~/workspace/jacobean/scion-sandbox-worker$ git push deis stateless-remote:master
Counting objects: 1, done.
Writing objects: 100% (1/1), 191 bytes | 0 bytes/s, done.
Total 1 (delta 0), reused 0 (delta 0)
-----> Node.js app detected
       
       Node engine:         unspecified
       Npm engine:          unspecified
       Start mechanism:     npm start
       node_modules source: package.json
       node_modules cached: true
       
       NPM_CONFIG_PRODUCTION=true
       NODE_MODULES_CACHE=true
       
-----> Installing binaries
       Resolving node version (latest stable) via semver.io...
       Downloading and installing node 0.12.2...
       
-----> Building dependencies
       Restoring node modules from cache
       Pruning unused dependencies
       Installing any new modules
       
-----> Checking startup method
       No Procfile; Adding 'web: npm start' to new Procfile
       
-----> Finalizing build
       Creating runtime environment
       Exporting binary paths
       Cleaning up build artifacts
       Caching node_modules for future builds
       
-----> Build successful!
       scion-sandbox-worker@0.0.0 /tmp/build
       ├── body-parser@1.10.2
       ├── cookie-parser@1.3.4
       ├── debug@2.1.3
       ├── express@4.11.2
       ├── jade@1.9.2
       ├── knox@0.9.2
       ├── morgan@1.5.2
       ├── redis@0.12.1
       ├── request@2.55.0
       ├── scxml@2.0.1
       ├── serve-favicon@2.2.0
       └── underscore@1.8.3
       
       WARNING: Node version not specified in package.json
       https://devcenter.heroku.com/articles/nodejs-support#specifying-a-node-js-version
       
-----> Discovering process types
       Procfile declares types -> web
-----> Compiled slug size is 12M

-----> Building Docker image
remote: Sending build context to Docker daemon 12.26 MB
remote: build context to Docker daemon 
Step 0 : FROM deis/slugrunner
 ---> 41e0b8aed6f7
Step 1 : RUN mkdir -p /app
 ---> Using cache
 ---> bd5fae69e696
Step 2 : WORKDIR /app
 ---> Using cache
 ---> 812837cbc247
Step 3 : ENTRYPOINT /runner/init
 ---> Using cache
 ---> e102c5333a6c
Step 4 : ADD slug.tgz /app
 ---> 0b1d7b6503c6
Removing intermediate container f9ef53d27c84
Step 5 : ENV GIT_SHA a1126b9f06dfbc0f000be0c0ff14521fcf7eedc4
 ---> Running in 7439c89cd13e
 ---> 134c163ecd44
Removing intermediate container 7439c89cd13e
Successfully built 134c163ecd44
-----> Pushing image to private registry

-----> Launching... 
       done, scion-sandbox-worker:v15 deployed to Deis

       http://scion-sandbox-worker.deis.scxml.io

       To learn more, use `deis help` or visit http://deis.io

To ssh://git@deis.scxml.io:2222/scion-sandbox-worker.git
   80e2ebf..a1126b9  stateless-remote -> master
```

You can see `NODE_MODULES_CACHE=true`

**Comment by aledbf** [comment]
@jbeard4 just as a suggestion (not related to the issue) use `npm shrinkwrap` and add the `npm-shrinkwrap.json` to git so you can reproduce the build in any environment.

**Comment by jbeard4** [comment]
@aledbf thanks for the tip. I will review this option

**Comment by mboersma** [comment]
The heroku-buildpack-nodejs is being updated to [v75](https://github.com/heroku/heroku-buildpack-nodejs/releases/tag/v75) in #3701, which may affect this issue.

**Comment by jbeard4** [comment]
@mboersma How would I go about testing this?

**Comment by bacongobbler** [comment]
Try `deis config:set BUIKDPACK_URL=https://github.com/heroku/heroku-buikdpack-nodejs#v75` before pushing your app

**Comment by jbeard4** [comment]
will do. thank you

**Comment by mboersma** [comment]
@jbeard4 did you have better results with the updated NodeJS buildpack?

**Comment by jbeard4** [comment]
@mboersma I will try it and let you know

**Comment by bacongobbler** [comment]
ping @jbeard4, any news? :)

**Comment by mboersma** [comment]
This issue is nine months old and likely fixed by #4976 or earlier. Please re-open if this problem persists in Deis v1.13.0.
